### ISSUE TITLE
clarification of featureParams

### ISSUE THREAD

**Issue by PeterConstable** [MicrosoftDocs/typography-issues#439]
In spite of current wording, some readers are having difficulty understanding that non-null offsets and description of feature parameter tables is in the descriptions of specific features. Might help to give a specific example.

Also, in the table describing the Feature table, the description for the featureParams field does say "= NULL (reserved for offset to FeatureParams)". (The examples at the end of the chapter also show "featureParams, reserved, null".) This is misleading as it suggests that there is no current usage for this field, which is not the case.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 1cb42967-20ca-9270-4572-65315127fcc5
* Version Independent ID: 1fa413f1-71cf-ac2a-0999-70dd5d8ac4cd
* Content: [OpenType layout common table formats - Typography](https://docs.microsoft.com/en-us/typography/opentype/spec/chapter2)
* Content Source: [typographydocs/opentype/spec/chapter2.md](https://github.com/MicrosoftDocs/typography/blob/live/typographydocs/opentype/spec/chapter2.md)
* Product: **typography**
* GitHub Login: @PeterCon
* Microsoft Alias: **PeterCon**

**Comment by Richard57** [comment]
There's also the sentence, "If a Feature Parameters table is not needed, the FeatureParams field must be set to NULL."  The word 'needed' should be replaced by something clearer, such as 'wanted', or 'wanted and defined', or perhaps weakened to 'to be supplied'.

**Comment by HinTak** [comment]
And an explicit reference to the current registered tags, somewhat along to line of "see registered tag section on details of individual feature tags for which FeatureParams are used..." , may be?

**Comment by HinTak** [comment]
The comments so far captured more adequately what I cannot quite express clearly. I have looked at the text again, and I think I would like to add two points:

- one "naturally" expects FeatureParams being part of the FeatureTable being part of the GSUB/GPOS tables, more details are where those tables are described, at least where an overview is, instead of at the festure tags (and their individual description and implementation) section.

- in where it says "null", "reserved", the surrounding paragraphs do not point to where it might not be - e.g. "...see the description and implementation section of individual feature tags for details..." would be nice?... And "reserved" definitely seems to sound like it is the "norm".

- I now understand the list of feature tags are (1) somewhat in flux and more are being added over time, (2) and somewhat vendor-specific too - vendor's software have better support to specific groups of tags vs others.

**Comment by HinTak** [comment]
> There's also the sentence, "If a Feature Parameters table is not needed, the FeatureParams field must be set to NULL." The word 'needed' should be replaced by something clearer, such as 'wanted', or 'wanted and defined', or perhaps weakened to 'to be supplied'.

I would reworded it as "If a Feature Parameters table is not needed for the specific feature tag, the FeatureParams field must be set to NULL. See the registered tag section on description and implementation details of individual tags." . The not-explicit part about "is not needed", is "under what circumstances is it needed or not needed?"

Hindsight is  wonderful thing :-) .

**Comment by Richard57** [comment]
That first sentence should be, "If a Feature Parameters table is not defined for the specific feature tag, the FeatureParams field must be set to NULL."  That avoids the issue of whether a table is an unnecessary luxury.

**Comment by PeterCon** [comment]
Changed for the next version:

>A Feature table consists of an offset to a Feature Parameters (FeatureParams) table<del> (if one has been defined for this feature - see note in the following paragraph)</del>, a count of the lookups listed for the feature (LookupCount), and an arbitrarily ordered array of indices into a LookupList (LookupListIndex). The LookupList indices are references into an array of offsets to Lookup tables.
>
> <ins>Feature Parameters tables can only be used for certain features.</ins> The format of the Feature Parameters table is specific to a particular feature, and must be specified in the feature’s entry in the Feature Tags section of the OpenType Layout Tag Registry. <ins>Currently, Feature Parameter tables are defined only for the following features:</ins>
>
> * <ins>[&#39;cv01&#39; – &#39;cv99&#39;](features_ae.md#cv01-cv99)</ins>
> * <ins>[&#39;size&#39;](features_pt.md#size)</ins>
> * <ins>[&#39;ss01&#39; – &#39;ss20&#39;](features_pt.md#ssxx)</ins>
>
> <ins>A Features Parameters table may be required or optional, according to the specifications for a given feature. </ins>The length of the Feature Parameters table must be implicitly or explicitly specified in the Feature Parameters table itself. The <del>FeatureParams</del><ins>featureParamsOffset</ins> field in the Feature Table <del>records</del><ins>gives</ins> the offset relative to the beginning of the Feature Table. If a Feature Parameters table is not <del>needed</del><ins>defined for a given feature, or if a Feature Parameters table is defined but not used in a given font</ins>, the <del>FeatureParams</del><ins>featureParamsOffset</ins> field must be set to NULL.

In the table describing the Feature table:

><table><tr><td>Offset16</td><td>featureParamsOffset</td><td><del>= NULL (reserved for offset to FeatureParams)</del><ins>Offset from start of Feature table to FeatureParams table, if defined for the feature and present, else NULL</ins></td></tr></table>

**Comment by HinTak** [comment]
Thanks a lot!
